### ISSUE TITLE
memoizing the setCookie and removeCookie functions returned by useCookie

### ISSUE THREAD

**Issue by justinhelmer** [reactivestack/cookies#217]
First I want to say thanks for building such an incredible tool, and the support for the new hooks API helps with building some truly beautiful components.

### Context
We recently added [eslint-plugin-react-hooks](https://www.npmjs.com/package/eslint-plugin-react-hooks), and are using the `exhaustive-deps` rule to help us see any optimization misses and make our code better. It has been incredibly helpful, but doesn't place nice with `react-cookie`.

For example, imagine that my functional component looks something like this:

```js
export default () => {
  const [cookies, , removeCookie] = useCookies(['uid']);
  const [isAuthenticated, setIsAuthenticated] = useState(typeof cookies.uid !== 'undefined');
  const [user, setUser] = useState({});
  const [userDataReady, setUserDataReady] = useState();

  const killUserSession = () => {
    removeCookie('uid');
    setUser({});
    setIsAuthenticated(false);
  });

  const setUserSession = (userJson) => {
    setUser(userJson);
    setIsAuthenticated(true);
  });

  useEffect(() => {
    if (isAuthenticated && !user.id) {
      // uid cookie exists but user state hasn't been populated
      const promise = fetch('/api/user')
        .then((userJson) => setUserSession(userJson))
        .catch(() => killUserSession()); // session expired

      setUserDataReady(promise);
    }
  });

  // return ... 
};
```

When using `eslint --fix`, the `react-hooks` plugin will brilliantly attempt to auto-optimize the code as per the `react-hooks/exhaustive-deps` rule. It will then become:

```js
export default () => {
  // omitting variible definitions for the sake of brevity

  const killUserSession = useCallback() => {
    removeCookie('uid');
    setUser({});
    setIsAuthenticated(false);
  }, [removeCookie]);

  const setUserSession = useCallback((userJson) => {
    setUser(userJson);
    setIsAuthenticated(true);
  }, []));

  useEffect(() => {
    if (isAuthenticated && !user.id) {
      // uid cookie exists but user state hasn't been populated
      const promise = fetch('/api/user')
        .then((userJson) => setUserSession(userJson))
        .catch(() => killUserSession()); // session expired

      setUserDataReady(promise);
    }
  }, [isAuthenticated, user.id, setUserSession, killUserSession]);

  // return ...
};
```

As you can see, the plugin wraps the function definitions in [useCallback](https://reactjs.org/docs/hooks-reference.html#usecallback), to generate memoized versions of them.

In terms of the dependencies to watch for changes, the plugin infers that `setUser` and `setAuthenticated` do not need to be in the dependency list for `setUserSession` or `killUserSession`. This  is probably because it knows that they come from `useState` and so they can infer that the definitions will not ever change. However, the plugin cannot determine whether or not `removeCookie` will change (that would be incredible if it could, but seems like it would be unreasonably complicated). For that reason, `killUserSession` includes `removeCookie` as a dependency to watch for changes, since that value is used internally by the memoized callback.

### Problem
This causes an infinite loop at execution-time.

1) After render, the `useEffect` side-effect is executed.
2) Immediately, the state is updated (i.e. `setUserDataReady` is called, which updates state).
3) Updating state will cause a re-render
4) After render the `useEffect` side-effect is run again, if any of the listed dependencies (`isAuthenticated`, `user.id`, `setUserSession`, `killUserSession`) change. Nothing will have changed except for `killUserSession`, which changes because `removeCookie` is listed as a dependency for the callback. Therefore the side-effect is fired again.
5) Because the `fetch()` operation is asynchronous, the user data will not yet be ready. Therefore the below condition will still be true:

```js
if (isAuthenticated && !user.id) {
```
...which causes the `fetch()` to happen again, and the state to update again. We circle back to step (1), ad-infinitum.

---

From this, we can infer that the `removeCookie` function is redefined every time `useCookie` is called. I tested that this behavior is the same when using the `setCookie` function.

### Solution
I am aware of the ideal solution of moving the function definitions for `setUserSession` and `killUserSession` inside the `useEffect` side-effect (as described [here](https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies)). In this case, it is not an option, because the callbacks are used elsewhere in the same component. I have just omitted the other usages for the sake of brevity.

I can work around this issue by omitting `removeCookie` from the list of dependencies to watch in the memoized `killUserSession` callback, and also disabling the lint rule for that line:

```js
const killUserSession = useCallback() => {
  removeCookie('uid');
  setUser({});
  setIsAuthenticated(false);
  // eslint-disable-next-line react-hooks/exhaustive-deps
}, []);
```

However, this is less-than-ideal, because disabling the lint rule entirely prevents us from gaining the benefits of the `react-hooks` plugin as we iterate and make changes on the function definition for `killUserSession`. Also, we use `react-cookie` in other places in the code, and therefore we would be losing the benefits in multiple places throughout the codebase.

Would it be possible to memoize the `setCookie` and `removeCookie` functions that are returned by the `useCookie` hook? By looking at the definition of [what is returned by useContext](https://github.com/reactivestack/cookies/blob/master/packages/react-cookie/src/useCookies.tsx#L48), I can see that you are using `.bind()`, which will always generate a new function.

Without digging much deeper in the source code or fully understanding the implications, what are your thoughts on using `useCallback` to prevent generating a new function definition every time? Something like:

```js
export default function useCookies() {
  // ...

  const setCookie = useMemo(() => cookies.set.bind(cookies), []);
  const removeCookie = useMemo(() => cookies.remove.bind(cookies), []);

  return [allCookies, cookies.set.bind(cookies), cookies.remove.bind(cookies)];
}
```

### References
- https://github.com/facebook/react/issues/14920#issuecomment-471070149 (look for "infinite loop")
- https://github.com/facebook/react/issues/15084
- https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

Thanks for your time!

**Comment by eXon** [comment]
Hey @justinhelmer 

Thank you for all the details, that is an amazing bug report 👍 .

I think it is a legitimate concern that should be addressed. The only missing part I see is the `cookies` instance as a dependency on the `useMemo`. It's not something that should change, but it is technically possible.

I'll dig a little deeper into the ressources you've provided. I haven't used the React hooks enough to completely understand the impacts but it feel important. Thank you! :)

**Comment by justinhelmer** [comment]
Thanks @eXon !

**Comment by eXon** [comment]
Fixed in `v4.0.0`!
